api/lib/opentelemetry/instrumentation/adapter.rb

```diff
@@ -0,0 +1,219 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    # The Adapter class holds all metadata and configuration for an
+    # instrumentation adapter. All instrumentation adapter packages should
+    # include a subclass of +Instrumentation::Adapter+ that will register
+    # it with +OpenTelemetry.instrumentation_regsitry+ and make it available for
+    # discovery and installation by an SDK.
+    #
+    # A typical subclass of Adapter will provide the adapter_name,
+    # adapter_version, an install block, a present block, and possibly
+    # a compatible block. Below is an example:
+    #
+    # module OpenTelemetry
+    #   module Adapters
+    #     module Sinatra
+    #       class Adapter < OpenTelemetry::Instrumentation::Adapter
+    #         library_name 'OpenTelemetry::Adapters::Sinatra'
+    #         library_version OpenTelemetry::Adapters::Sinatra::VERSION
```

Comment:
Can we rely on conventions to avoid having to specify the `library_name` and `library_version`? I see you have defaults for both, but in this case they would be equivalent to:
```ruby
library_name 'OpenTelemetry::Adapters::Sinatra::Adapter'
library_version '0.0.0'
```

It feels like we could make more assumptions here, and 💥 in informative ways if the assumptions are invalid.

Reply:
Also, I think you renamed these `adapter_name` and `adapter_version`

Reply:
I added the ability to infer name and version. The name is inferred as the namespace if the class is namespaced, otherwise it is the class name. On the off chance the class name is nil (this would happen for an anonymous class) it will be 'unknown'.

Version is similar. It defaults to the `VERSION` constant under the namespace if it exists, the class if it doesn't, and '0.0.0` if neither exist.

I'm trying to avoid :boom:ing if I can since the defaults are lazily evaluated when the instance is created.